code/game/objects/items/mail.dm

```diff
@@ -230,17 +251,22 @@
 
 	//better spam mail names instead of being "IMPORTANT MAIL", courtesy of Monkestation
 	color = "#[pick(random_short_color())]"
-	switch(rand(1,10))
-
-		if(1,2)
-			name = special_name ? junk_names[junk] : "[initial(name)] for [pick(GLOB.alive_mob_list)]" //LETTER FOR IAN / BUBBLEGUM / MONKEY(420)
-		if(3,4)
-			name = special_name ? junk_names[junk] : "[initial(name)] for [pick(GLOB.player_list)]" //Letter for ANYONE, even that wizard rampaging through the station.
-		if(5)
-			name = special_name ? junk_names[junk] : "DO NOT OPEN"
-		else
-			name = special_name ? junk_names[junk] : "[pick("important","critical","crucial","serious","vital")] [initial(name)]"
-
+	if(recipient)
+		name = "[initial(name)] for [recipient.name] ([recipient.assigned_role])"
+		recipient_ref = WEAKREF(recipient)
+	else
+		switch(rand(1,10))
+			/*
+			if(1,2)
+				name = special_name ? junk_names[junk] : "[initial(name)] for [pick(GLOB.alive_mob_list)]" //LETTER FOR IAN / BUBBLEGUM / MONKEY(420)
+			*/ //NSV13 Having a lot of trouble with mail being generated for entities the crew aren't supposed to know about
+			if(1,2,3,4)//NSV13 added 1 and 2
+				name = special_name ? junk_names[junk] : "[initial(name)] for [pick(GLOB.player_list)]" //Letter for ANYONE, even that wizard rampaging through the station.
```

Comment:
This is probably why ghost mail is happening, because ghosts are also players. To get what we want, I would break it up so that generating crew spam and generating simplemob spam are different numbers. 1,2 and 3,4 or 1,2,3 and 4, depending on what distribution we want.
I think picking a **crew member** regardless of Z level (so miners and pilots can have mail) looks like this:
```
var/datum/data/record/t = pick(GLOB.data_core.general)
name = special_name ? junk_names[junk] : "[initial(name)] for [t.fields["name"]]"
```
Picking a simplemob on the ship is more complicated, but I would probably use the NPC list instead of the alive mobs list:
```
var/list/candidates = list()
for(var/status in GLOB.simple_animals)
	for(var/mob/simple_animal/M in GLOB.simple_animals[status])
		if(level_trait(M.z, ZTRAIT_STATION) // Ignore anything not on the main ship
			candidates += M
name = special_name ? junk_names[junk] : "[initial(name)] for [pick(candidates)]"
```
This is admittedly a bit expensive. We could cache the first candidate list we generate, but this method would not accommodate any mobs added after the first time we created spam mail:

```
var/static/list/candidates
if(!candidates)
	var/list/candidates = list()
	for(var/status in GLOB.simple_animals)
		for(var/mob/simple_animal/M in GLOB.simple_animals[status])
			if(level_trait(M.z, ZTRAIT_STATION)
				candidates += M
name = special_name ? junk_names[junk] : "[initial(name)] for [pick(candidates)]"
```